src/mpi/coll/transports/generic/transport.c

```diff
@@ -0,0 +1,1243 @@
+
+    /* store the arguments */
+    vtxp->nbargs.dtcopy.tobuf = tobuf;
+    vtxp->nbargs.dtcopy.tocount = tocount;
+    vtxp->nbargs.dtcopy.totype = totype;
+    vtxp->nbargs.dtcopy.frombuf = frombuf;
+    vtxp->nbargs.dtcopy.fromcount = fromcount;
+    vtxp->nbargs.dtcopy.fromtype = fromtype;
+
+    MPL_DBG_MSG_FMT(MPIR_DBG_COLL,VERBOSE,(MPL_DBG_FDEST,"TSP(generic) : sched [%d] [dt_copy]\n", vtx_id));
+    return vtx_id;
+}
+
+/* Internal transport function to add a single element to MPIR_COLL_GENERIC_ptr_array */
+MPL_STATIC_INLINE_PREFIX void MPIR_COLL_GENERIC_add_elem_ptr_array(MPIR_COLL_GENERIC_ptr_array * buf_array, void *buf)
+{
+    if (buf_array->used == buf_array->size) {   /* increase array size */
+        int old_size = buf_array->size;
+        void **old_array = buf_array->array;
+        buf_array->size *= 2;
+        /* allocate new array */
+        buf_array->array = (void **) MPIR_COLL_GENERIC_allocate_mem(sizeof(void *) * buf_array->size);
+        /* copy from old array to new array */
+        memcpy(buf_array->array, old_array, sizeof(void *) * old_size);
+        /* free old array */
+        MPIR_COLL_GENERIC_free_mem(old_array);
+    }
+    /* add element to the array */
+    buf_array->array[buf_array->used++] = buf;
+}
+
+/* Transport function to allocate memory required for schedule execution.
+ *The memory address is recorded by the transport in the schedule
+ * so that this memory can be freed when the schedule is destroyed
+ */
+void *MPIR_COLL_GENERIC_allocate_buffer(size_t size, MPIR_COLL_GENERIC_sched_t * s)
+{
+    void *buf = MPIR_COLL_GENERIC_allocate_mem(size);
+    /* record memory allocation */
+    MPIR_COLL_GENERIC_add_elem_ptr_array(&s->buf_array, buf);
+    return buf;
+}
+
+/* Transport function to schedule memory free'ing */
+int MPIR_COLL_GENERIC_free_mem_nb(void *ptr,
+                                       MPIR_COLL_GENERIC_sched_t * sched, int n_invtcs, int *invtcs)
+{
+    MPIR_COLL_GENERIC_vtx_t *vtxp;
+    int vtx_id = MPIR_COLL_GENERIC_get_new_vtx(sched, &vtxp);
+
+    /* assign a new vertex */
+    vtxp->kind = MPIR_COLL_GENERIC_KIND_FREE_MEM;
+    MPIR_COLL_GENERIC_init_vtx(sched, vtxp, vtx_id);
+    MPIR_COLL_GENERIC_add_vtx_dependencies(sched, vtx_id, n_invtcs, invtcs);
+
+    /* store the arguments */
+    vtxp->nbargs.free_mem.ptr = ptr;
+
+    MPL_DBG_MSG_FMT(MPIR_DBG_COLL,VERBOSE,(MPL_DBG_FDEST,"TSP(generic) : sched [%d] [free_mem]\n", vtx_id));
+    return vtx_id;
+}
+
+#undef FUNCNAME
+#define FUNCNAME MPIR_COLL_GENERIC_issue_vtx
+/* Internal transport function to issue a vertex */
+/* This is set to be static inline because compiler cannot build this function with MPL_STATIC_INLINE_PREFIX
+ * complaining that this is a non-inlineable function */
+static inline void MPIR_COLL_GENERIC_issue_vtx(int vtxid, MPIR_COLL_GENERIC_vtx_t * vtxp,
+                                        MPIR_COLL_GENERIC_sched_t * sched)
```

Comment:
Remove namespace (or move it to another file and make it an internal-only namespace like `MPII_COLL_GENERIC_issue_vtx`).

Reply:
Why is this function non-inlineable?

Reply:
I don't know. A couple of months ago, the compiler was complaining and giving error. Need to check if it still persists.

Reply:
Yeah, I don't like seeing these weird hacks without a good reason. We should figure it out and document it if we can't fix it.